### PR TITLE
Add performance testing framework: preparatory refactoring

### PR DESCRIPTION
**Pull request by tamiko**
@peterrum and I are currently working on bringing back performance tests to deal.II. Our idea is to integrate performance tests as a part of the testsuite with the following rationale:
 - By default performance tests will be compiled and linked but do <b>not</b> run (with expected stage `RUN` or `DIFF`). This will ensure that compilation and linking of performance test code is part of our CI and does not diverge from the library API.
 - We will use `ctest` (and our testsuite infrastructure) to compile and run performance tests that simply report (a best of three) timing result to the output file. These output files can then further be postprocessed by a script to create a table. All of this will be chained together in a `run_performance_tests.cmake` ctest script.

This PR adds some necessary features to the current testsuite infrastructure:
 - [x] implement thread concurrency annotation via `[...].threads=XY.output`
 - [x] the ability to set MPI rank and thread limits via `TEST_THREAD_LIMIT` and `TEST_MPI_RANK_LIMIT`. Tests that specify a higher number of MPI ranks or thread concurrency will be dropped.
 - [x] allow to use maximal concurrency by setting `.threads=max.` or `.mpirun=max.`.

### CODE REVIEW

File: cmake/macros/macro_deal_ii_add_test.cmake [hunk 1]
```diff
@@ -391,9 +462,9 @@ MACRO(DEAL_II_ADD_TEST _category _test_name _comparison_file)
             DEPENDS ${TEST_DEPENDENCIES_${_target}}
             )
         ENDIF()
-        SET(TEST_DEPENDENCIES_${_target} ${_test_full})
+        SET(TEST_DEPENDENCIES_${_target} ${_test_full} PARENT_SCOPE)
```
Comment (peterrum) [inline cmake/macros/macro_deal_ii_add_test.cmake:465]: Why did you add `PARENT_SCOPE`?

Comment (tamiko) [reply cmake/macros/macro_deal_ii_add_test.cmake:465]: Because I converted the macro into a function in order to use `RETURN()`.
In order to set a variable name in the (now) parent scope I have to use the `PARENT_SCOPE` keyword.

**Review by peterrum — COMMENTED** [review]
Looks good to me! Thanks for making the first steps!

Since I am too much involved here, I would appreciate a second opinion!

**Review by masterleinad — APPROVED** [review]
Looks good to me.

**Review by peterrum — APPROVED** [review]
Let's merge. More is comming!
